### ISSUE TITLE
need usage help

### ISSUE THREAD

**Issue by lakamsani** [mevansam/terraform-provider-cf#30]
Hi,

This is a bit long, please bear with me as I get familiar with this provider and hopefully provide some useful testing feedbck if it works out.

We have our app setup in a dev  Bluemix/IBM Cloud  env manually using redis, postgres, a CUPS that stores encrypted passwords for the two dbs that the apps read at runtime, auto-scaling and dynatrace. We need to setup a few more envs similar to this (say test, production) and trying to figure out if it makes sense to use this provider. So far I 'm able to connect and print org id, space ids etc. (thanks to answers provided for #29)

This is the high level plan:

1.  Use this provider to create various services with pre-determined names (the app's manifest.xml file will use the same names to bind when app is deployed via cf push). In the case of db services find a way to get the user host/port/user id/passwords into Terraform output variables.
2.  Run some shell commands from Terraform  (https://www.terraform.io/docs/provisioners/local-exec.html ) to:
- (a) update the db host/port/userid in the app's manifest.xml and 
- (b) separate curl command store the encrypted database passwords in the CUPS. App will reference them via place holders in manifest.xml and populate them at run time by talking to the CUPS.

 Will this work? Don't understand the purpose of `cf_app` resource in this provider given that the app can be deployed initially and subsequent times via `cf push`. Is `cf_app` an  alternative way of deploying/re-deploying a CF app but using this provider?

And finally, I 'm not able to figure out how to create a new redis and postgres services.  I think I should start with `cf_service_instance`. On its wiki page a couple of links under the Argument Reference section are broken. 

https://github.com/mevansam/terraform-provider-cf/blob/master/website/docs/r/service_instance.html.markdown

Our Service Plans for redis and postgres are Enterprise, for auto scaling its Dedicated and for dynatrace is standard. Can I just hard code them when creating those service instances or do I need to read the redis service first using `data` and pass the id of the Enterprise service plan by pulling it from service_plans array using something like `${data.cf_service.redis.service_plans["enterprise"]}`  as shown on the `service_instance` wiki page linked above?

I first tried with 

```bash
data "cf_service" "redis" {
  name = "Compose for Redis"
}
```
but got this error
```bash
Error: Error refreshing state: 1 error(s) occurred:

* data.cf_service.redis: 1 error(s) occurred:

* data.cf_service.redis: data.cf_service.redis: Service Compose for Redis not found
```

We have a service catalog that shows Compose for Redis, Dedicated and Compose for Postgres, Dedicated. How can I get this `data` statement work as I think this is needed to get `service_instance` to work. 

I tried by supply the string "Dedicated" to the `service_plan` parameter of the `service_instance` resource. `terraform plan` didn't give any error and printed this:

```bash
Terraform will perform the following actions:

  + cf_service_instance.redis
      id:           <computed>
      name:         "redis"
      service_plan: "Dedicated"
      space:        "4de30f7a-9c62-4034-89cf-d5a362ba1ebc"
```

But when I ran `terraform apply` it gave this error:

```bash
* cf_service_instance.redis2: Server error, status code: 400, error code: 60003, message: The service instance is invalid: not a valid service plan
```

**Comment by gberche-orange** [comment]
@lakamsani 

> I first tried with
> 
> data "cf_service" "redis" {
>   name = "Compose for Redis"
> }
> 
> but got this error
> 
> Error: Error refreshing state: 1 error(s) occurred:
> 
> * data.cf_service.redis: 1 error(s) occurred:
> 
> * data.cf_service.redis: data.cf_service.redis: Service Compose for Redis not found

I suspect the problem is that the specified service name `Service Compose for Redis` does not match the service catalog. On bluemix when using `cf marketplace` you get the following output

```
cf m
OK

service                                       plans                                                                                                                                                                                                                                                                                                                                                                                           description
[...]
compose-for-redis                             Standard*, Enterprise*                                                                                                                                                                                                                                                                                                                                                                          Redis is an open-source, blazingly fast, low maintenance key/value store.
```

Try specifying `compose-for-redis` instead of the bluemix-ui-specific `Service Compose for Redis`

> I tried by supply the string "Dedicated" to the service_plan parameter of the service_instance resource. terraform plan didn't give any error and printed this:

[cf_service_instance.service_plan](https://github.com/mevansam/terraform-provider-cf/blob/master/website/docs/r/service_instance.html.markdown) expects a guid of the service plan.

As an alternative to the data source syntax `"${data.cf_service.redis.service_plans["shared-vm"]}"` you may use the CF CLI to discover a service plan:

```
CF_TRACE=true cf m -s compose-for-redis | grep guid
        "guid": "d48f8c52-2987-4cc0-ad60-a3c633c3fce7",
        "service_guid": "c6eb6f8a-3501-4fdb-b41e-d9ccdd713a07",
        "guid": "7f061c00-e04e-4d4d-a766-0ab3ab9081ff",
        "service_guid": "c6eb6f8a-3501-4fdb-b41e-d9ccdd713a07",
```


> Don't understand the purpose of cf_app resource in this provider given that the app can be deployed initially and subsequent times via cf push. Is cf_app an alternative way of deploying/re-deploying a CF app but using this provider?

Yes using this provider, you should be able to craft a TF module that enables you to deploy a given app (along with prereq services) in multiple spaces and have them be consistent among each other (parity between dev/test/production)

ps:  we prefer support requests to be addressed to https://cloudfoundry.slack.com/messages/C7JRBR8CV/team/U0LRV26PN/ and prefer github issues for bugs or enhancement requests. Please reopen this issue if you believe there is bug or would like to suggest improvements

**Comment by lakamsani** [comment]
@gberche-orange  thanks for the suggestions and tip about Slack group for this provider.

I 'm a bit more familiar with AWS than CF and didn't know about `cf m` at all. I was able to get redis instance created in my environment by using the first guid which is under `metadata.guid` instead of `entity.service_guid` from the output of the command: `CF_TRACE=true cf m -s compose-redis-dedicated`. 

After that I tried to figure out how to get it from `data` and found that this seems to work. ${data.cf_service.redis.service_plans["Enterprise"]}.  OTOH `${data.cf_service.redis.id}`   maps to `entity.service_guid` which is no good for creating the redis service instance.
